syntax = "proto3";

package product;

service ProductService {
  rpc CreateProduct (CreateProductRequest) returns (CreateProductResponse);
  rpc GetProduct (GetProductRequest) returns (GetProductResponse);
  rpc GetProducts (GetAllProductsRequest) returns (GetAllProductsResponse);  
  rpc ValidateSkuInputs (ValidateSkuInputRequest) returns (ValidateSkuInputResponse);
  rpc UpdateProduct (UpdateProductRequest) returns (UpdateProductResponse);
  rpc DeleteProduct (DeleteProductRequest) returns (DeleteProductResponse);
  rpc existingSku (CheckSkuAvailabilityRequest) returns (CheckSkuAvailabilityResponse);
  rpc GetAttributes (GetAttributesRequest) returns (GetAttributesResponse);
  rpc GetAttributeOptions (GetAttributeOptionsRequest) returns (GetAttributeOptionsResponse);
}

message GetAttributesRequest {}

message GetAttributesResponse {
  repeated AttributeDetail attributes = 1;
}

message GetAttributeOptionsRequest {
  string attributeId = 1;
}

message GetAttributeOptionsResponse {
  string attributeId = 1;
  string attributeName = 2;
  repeated AttributeOption options = 3;
}

message AttributeOption {
  string id = 1;
  string value = 2;
  string description = 3;
}

message AttributeDetail {
  string id = 1;
  string name = 2;
  string description = 3;
}

enum SortField {
  SORT_FIELD_UNSPECIFIED = 0;
  SORT_FIELD_NAME = 1;
  SORT_FIELD_PRICE = 2;
  SORT_FIELD_CREATED_AT = 3;
}

enum SortOrder {
  SORT_ORDER_UNSPECIFIED = 0;
  SORT_ORDER_ASC = 1;
  SORT_ORDER_DESC = 2;
}

message BrandResponse {
  string id = 1;
  string name = 2;
}

message CategoryResponse {
  string id = 1;
  string name = 2;
  string slug = 3;
}

message SkuResponse {
  string id = 1;
  string skuCode = 2;
  double price = 3;
  int32 stock = 4;
  string image = 5;
  repeated SkuOptionResponse skuOptions = 6;
  double originalPrice = 7;
}

message SkuOptionResponse {
  string attributeOptionId = 1;
  string attributeOptionValue = 2;
  AttributeDetail attribute = 3;
}

message CreateProductRequest {
  string name = 1;
  string description = 2;
  string brandId = 3;
  string categoryId = 4;
  repeated CreateSkuInput skus = 5;
  double originalPrice = 6;
}

message CreateSkuInput {
  string skuCode = 1;
  double price = 2;
  int32 stock = 3;
  string image = 4;
  repeated AttributeInput attributes = 5;
}

message AttributeInput {
  string attributeOptionId = 1;
}

message CreateProductResponse {
  GetProductResponse product = 1;
}

message GetProductRequest {
  string id = 1;
  optional string userId = 2;
}

message GetProductResponse {
  string id = 1;
  string name = 2;
  string description = 3;
  BrandResponse brand = 4;
  CategoryResponse category = 5;
  repeated SkuResponse skus = 6;
  double originalPrice = 7;
}

message GetProductSkuRequest {
  string id = 1;
}

message GetProductSkuResponse {
  string id = 1;
  string name = 2;
  string description = 3;
  BrandResponse brand = 4;
  CategoryResponse category = 5;
  repeated SkuResponse skus = 6;
}

message UpdateProductRequest {
  string id = 1;
  string name = 2;
  string description = 3;
  string brandId = 4;
  string categoryId = 5;
  repeated CreateSkuInput skus = 6;
  double originalPrice = 7;
}

message UpdateProductResponse {
  GetProductResponse product = 1;
}

message DeleteProductRequest {
  string id = 1;
}

message DeleteProductResponse {
  bool message = 1;
}

message GetAllProductsRequest {
  int32 page = 1;
  int32 limit = 2;
  string keyword = 3;
  string brandId = 4;
  string categoryId = 5;
  SortField sortField = 6;
  SortOrder sortOrder = 7;
  optional double minPrice = 8;
  optional double maxPrice = 9;
}

message GetAllProductsResponse {
  repeated GetProductResponse products = 1;
  int32 total = 2;
  int32 page = 3;        
  int32 limit = 4;       
  int32 totalPages = 5;  
}

message SkuValidationInput {
  string productId = 1;
  string skuId = 2;
  string skuCode = 3;
  int32 quantity = 4;
}

message ValidateSkuInputRequest {
  repeated SkuValidationInput items = 1;
}

message SkuValidationResult {
  string productId = 1;
  string name = 2;
  string description = 3;
  BrandResponse brand = 4;
  CategoryResponse category = 5;

  string skuId = 6;
  string skuCode = 7;
  double price = 8;
  int32 stock = 9;
  string image = 10;
  repeated SkuOptionResponse skuOptions = 11;
  bool valid = 12;
  bool inStock = 13;
  int32 availableStock = 14;
  string message = 15;
}

message ValidateSkuInputResponse {
  bool allValid = 1;
  repeated SkuValidationResult results = 2;
}

// ======================= SKU Availability =======================
message CheckSkuAvailabilityRequest {
  string skuId = 1;      
  int32 quantity = 2;     
}

message CheckSkuAvailabilityResponse {
  bool exists = 1;
  bool inStock = 2;
  int32 availableStock = 3;
  string message = 4;
  string productId = 5;
  string skuCode = 6;
  double price = 7;
  string productName = 8;
  string description = 9;
  string image = 10;
  BrandResponse brand = 11;
  CategoryResponse category = 12;
  repeated SkuOptionResponse skuOptions = 13;
}
